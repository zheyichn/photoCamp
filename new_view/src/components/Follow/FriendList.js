import React, { useState, useEffect } from "react";
import "../../styles/frendlistStyle.css";
import {
  followUserAPI,
  unfollowUserAPI,
} from "../../api/mock_api_following_post_follower";

export default function FriendList(props) {
  const [friendlist, setFriends] = useState([]);

  useEffect(() => {
    setFriends(props.friends);
  });

  function toggleFollowUnfollow(id) {
    const newfriendlist = [...friendlist];
    const friend = newfriendlist.find((friend) => friend._id === id);

    // update in API;
    if (friend.isFollow === false) {
      followUserAPI(sessionStorage.userName, friend.userName);
    } else {
      unfollowUserAPI(sessionStorage.userName, friend.userName);
    }

    friend.isFollow = !friend.isFollow;
    setFriends(newfriendlist);
  }

  function handleFollow(event) {
    toggleFollowUnfollow(event.target.value);
  }

  function createFriendBlock(imageUrl, name, id, isFollow) {
    return (
      <div className="friend-list-item" key={id}>
        <img className="item-photo" src={imageUrl} />
        <div className="item-info">
          <div className="item-name">{name}</div>
        </div>
        <button
          className="follow-following-btn"
          onClick={handleFollow}
          value={id}
        >
          {isFollow ? "Following" : "Follow"}
        </button>
      </div>
    );
  }

  function getFriendsBlocks(friendlist) {
    let friendsBlocks = [];
    for (let i = 0; i < friendlist.length; i++) {
      friendsBlocks.push(
        createFriendBlock(
          friendlist[i].profile,
          friendlist[i].userName,
          friendlist[i]._id,
          friendlist[i].isFollow
        )
      );
    }
    return friendsBlocks;
  }

  return (
    <div className="friend-list">
      <div>{getFriendsBlocks(friendlist)}</div>
    </div>
  );
}
