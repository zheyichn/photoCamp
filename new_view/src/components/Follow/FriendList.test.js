import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import FriendList from "./FriendList";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});

const friends = [
  {
    userID: 11,
    userName: "Julian Bartell DVM",
    profilePhoto: "http://loremflickr.com/640/480/business",
    isFollow: true,
    id: "1",
  },
  {
    userID: 56,
    userName: "Felix Christiansen",
    profilePhoto: "http://loremflickr.com/640/480/business",
    isFollow: true,
    id: "2",
  },
  {
    userID: 94,
    userName: "Edna Yost",
    profilePhoto: "http://loremflickr.com/640/480/transport",
    isFollow: false,
    id: "3",
  },
  {
    userID: 52,
    userName: "Faith Willms",
    profilePhoto: "http://loremflickr.com/640/480/technics",
    isFollow: false,
    id: "4",
  },
  {
    userID: 16,
    userName: "Rodolfo Lubowitz",
    profilePhoto: "http://loremflickr.com/640/480/business",
    isFollow: false,
    id: "5",
  },
  {
    userID: 39,
    userName: "Ida Lind Jr.",
    profilePhoto: "http://loremflickr.com/640/480/technics",
    isFollow: true,
    id: "6",
  },
];

describe("test friendlist page", () => {
  it("Create friendlist page body renders correctly", async () => {
    const tree = renderer.create(<FriendList friends={friends} />).toJSON();
    await expect(tree).toMatchSnapshot();
  });
});
