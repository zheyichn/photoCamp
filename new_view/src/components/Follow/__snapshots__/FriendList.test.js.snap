// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test friendlist page Create friendlist page body renders correctly 1`] = `
<div
  className="friend-list"
>
  <div />
</div>
`;
